src/state/image-dataset/csv-dataset/index.ts

```diff
@@ -0,0 +1,398 @@
+} from "../../metadata/types";
+import { ImageDataset, InitialDatasetSelections, Megaset } from "../types";
+import firebase from "firebase";
+import {
+    CELL_ID_KEY,
+    FOV_ID_KEY,
+    FOV_THUMBNAIL_PATH,
+    FOV_VOLUME_VIEWER_PATH,
+    GROUP_BY_KEY,
+    THUMBNAIL_PATH,
+    TRANSFORM,
+    VOLUME_VIEWER_PATH,
+} from "../../../constants";
+
+export const DEFAULT_CSV_DATASET_KEY = "csv";
+
+const METADATA_KEYS = new Set([
+    CELL_ID_KEY,
+    FOV_ID_KEY,
+    FOV_THUMBNAIL_PATH,
+    FOV_VOLUME_VIEWER_PATH,
+    THUMBNAIL_PATH,
+    VOLUME_VIEWER_PATH,
+    TRANSFORM,
+]);
+
+// From Adobe categorical colors
+const DEFAULT_COLORS = [
+    "#27B4AE",
+    "#4047C4",
+    "#F48730",
+    "#DB4281",
+    "#7E84F4",
+    "#78DF76",
+    "#1C7AED",
+    "#7129CD",
+    "#E7C73B",
+    "#C95F1E",
+    "#188E61",
+    "#BEE952",
+];
+
+function isNumeric(value: string): boolean {
+    if (typeof value != "string") {
+        return false;
+    }
+    if (value.trim().toLowerCase() === "nan") {
+        return true;
+    }
+    return !isNaN(Number(value)) && !isNaN(parseFloat(value));
+}
+
+function isStringArray(data: string[] | (number | null)[]): data is string[] {
+    return data.length > 0 && typeof data[0] === "string";
+}
+
+const enum FeatureType {
+    CONTINUOUS,
+    DISCRETE,
+}
+
+type FeatureInfo =
+    | {
+          type: FeatureType.CONTINUOUS;
+          def: ContinuousMeasuredFeatureDef;
+          data: (number | null)[];
+      }
+    | {
+          type: FeatureType.DISCRETE;
+          def: DiscreteMeasuredFeatureDef;
+          data: (number | null)[];
+      };
+
+/**
+ * Parses and mocks an ImageDataset from a provided CSV string with a header row.
+ *
+ * The CSV must contain:
+ * - URL paths to volume data, under the column name "volumeViewerPath"
+ *
+ * Optionally, the CSV can contain:
+ * - Thumbnail paths, under the column name "thumbnailPath"
+ * - Cell IDs, under the column name "CellId"
+ * - FOV IDs, under the column name "FOVId"
+ * - FOV thumbnail paths, under the column name "fovThumbnailPath"
+ * - FOV volume data, under the column name "fovVolumeviewerPath"
+ *
+ * Some data will be ignored by default:
+ * - Transform data, under the column name "transform"
+ *
+ * Any other columns will be interpreted as features:
+ * - Columns containing only numbers will be treated as numeric data.
+ * - Columns containing any non-numeric data will be treated as category ("discrete") data.
+ */
+class CsvRequest implements ImageDataset {
+    csvData: Record<string, string>[];
+    idToIndex: Record<string, number>;
+    featureInfo: Map<string, FeatureInfo>;
+
+    defaultGroupByFeatureKey: string;
+
+    constructor(csvFileContents: string) {
+        this.csvData = [];
+        this.idToIndex = {};
+        this.featureInfo = new Map();
+        // TODO: Automatically detect a discrete feature and replace the group
+        // by feature with it.
+        this.defaultGroupByFeatureKey = "CellId";
```

Comment:
Review comment seemed to get deleted here. Just asking if you can use a the CELL_ID_KEY constant here?